File: spec/javascript/test_repeat_until_complex.js
var assert = require('assert');
var testHelper = require('testHelper');

testHelper('RepeatUntilComplex', 'src/repeat_until_complex.bin', function(r) {
  assert.equal(r.first.length, 3);
  assert.equal(r.first[0].count, 4);
  assert.deepEqual(r.first[0].values, [1, 2, 3, 4]);
  assert.equal(r.first[1].count, 2);
  assert.deepEqual(r.first[1].values, [1, 2]);
  assert.equal(r.first[2].count, 0);
  assert.deepEqual(r.first[2].values, []);

  assert.equal(r.second.length, 4);
  assert.equal(r.second[0].count, 6);
  assert.deepEqual(r.second[0].values, [1, 2, 3, 4, 5, 6]);
  assert.equal(r.second[1].count, 3);
  assert.deepEqual(r.second[1].values, [1, 2, 3]);
  assert.equal(r.second[2].count, 4);
  assert.deepEqual(r.second[2].values, [1, 2, 3, 4]);
  assert.equal(r.second[3].count, 0);
  assert.deepEqual(r.second[3].values, []);
  
  assert.deepEqual(r.third, [102, 111, 111, 98, 97, 114, 0]);
});

File: spec/java/src/io/kaitai/struct/spec/TestRepeatUntilComplex.java
package io.kaitai.struct.spec;

import io.kaitai.struct.testformats.RepeatUntilComplex;
import org.testng.annotations.Test;

import static org.testng.Assert.assertEquals;

public class TestRepeatUntilComplex extends CommonSpec {
    @Test
    public void testRepeatUntilComplex() throws Exception {
        RepeatUntilComplex r = RepeatUntilComplex.fromFile(SRC_DIR + "repeat_until_complex.bin");

        assertEquals(r.first().size(), 3);
        assertEquals(r.first().get(0).count(), 4);
        assertEquals(r.first().get(0).values().toArray(), new Integer[] { 1, 2, 3, 4 });
        assertEquals(r.first().get(1).count(), 2);
        assertEquals(r.first().get(1).values().toArray(), new Integer[] { 1, 2 });
        assertEquals(r.first().get(2).count(), 0);
        assertEquals(r.first().get(2).values().toArray(), new Integer[] {  });
        assertEquals(r.second().size(), 4);
        assertEquals(r.second().get(0).count(), 6);
        assertEquals(r.second().get(0).values().toArray(), new Integer[] { 1, 2, 3, 4, 5, 6 });
        assertEquals(r.second().get(1).count(), 3);
        assertEquals(r.second().get(1).values().toArray(), new Integer[] { 1, 2, 3 });
        assertEquals(r.second().get(2).count(), 4);
        assertEquals(r.second().get(2).values().toArray(), new Integer[] { 1, 2, 3, 4 });
        assertEquals(r.second().get(3).count(), 0);
        assertEquals(r.second().get(3).values().toArray(), new Integer[] {  });
        assertEquals(r.third().toArray(), new Integer[] { 102, 111, 111, 98, 97, 114, 0 });
    }
}
